### PR TITLE
Convert BoundaryCondition to factory map

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Converts the system specific BoundaryCondition base classes to using the factory map.  

### Upgrade instructions

- Instead of having a type alias of `creatable_classes` within the system-specific BoundaryCondition base classes, specify a type alias such as `standard_boundary_conditions` in the BoundaryConditions namespace of the system.  
- In evolution executables, add a `tmpl::pair<SYSTEM::BoundaryConditions::BoundaryCondition, SYSTEM::BoundaryConditions::standard_boundary_conditions>` to the `factory_classes` in the metavariables (where SYSTEM should be replaced by the appropriate namespace of the evolution system).
- You don't need the system-specific `BoundaryConditions::register_derived_with_charm` function in `charm_init_node_funcs` anymore.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

This will allow #3468 to not need anything related to an analytic solution

### CODE REVIEW

File: tests/Unit/Evolution/Systems/Burgers/BoundaryConditions/Test_Outflow.cpp [hunk 1]
```diff
@@ -26,10 +26,6 @@ SPECTRE_TEST_CASE("Unit.Burgers.BoundaryConditions.Outflow",
   pypp::SetupLocalPythonEnvironment local_python_env{
       "Evolution/Systems/Burgers/BoundaryConditions/"};
   MAKE_GENERATOR(gen);
-  const std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>
-      boundary_condition = TestHelpers::test_creation<
-          std::unique_ptr<Burgers::BoundaryConditions::BoundaryCondition>>(
-          "Outflow:\n");
```
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/Burgers/BoundaryConditions/Test_Outflow.cpp:32]: Why did you delete this?

File: tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Test_Outflow.cpp [hunk 2]
```diff
@@ -26,10 +26,6 @@ SPECTRE_TEST_CASE("Unit.GrMhd.BoundaryConditions.Outflow", "[Unit][GrMhd]") {
   pypp::SetupLocalPythonEnvironment local_python_env{
       "Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/"};
   MAKE_GENERATOR(gen);
-  const std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>
-      boundary_condition = TestHelpers::test_creation<std::unique_ptr<
-          grmhd::ValenciaDivClean::BoundaryConditions::BoundaryCondition>>(
-          "Outflow:\n");
```
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Test_Outflow.cpp:32]: Same here

**Review by nilsvu — CHANGES_REQUESTED** [review]
Looks good. Have you considered deleting the system-specific base classes and just use `domain::BoundaryConditions::BoundaryCondition` as base class everywhere? That removes a lot of boilerplate code and some templating.

File: tests/Unit/Evolution/Systems/Burgers/BoundaryConditions/Test_Outflow.cpp (hunk 1 above)
Comment (kidder) [reply tests/Unit/Evolution/Systems/Burgers/BoundaryConditions/Test_Outflow.cpp:32]: It is tested inside`helpers::test_boundary_condition_with_python`

**Comment by kidder** [comment]
> Looks good. Have you considered deleting the system-specific base classes and just use `domain::BoundaryConditions::BoundaryCondition` as base class everywhere? That removes a lot of boilerplate code and some templating.

Yes, I am considering that

**Comment by nilsvu** [comment]
Needs a rebase to accommodate #3407

**Comment by wthrowe** [comment]
This looks good, once it is rebased.

**Comment by kidder** [comment]
@nilsleiffischer updated CurvedScalarWave and rebased

**Review by nilsvu — APPROVED** [review]

**Comment by nilsvu** [comment]
Something went wrong with CI here. Maybe rebase/push again?

**Pull request merged**